zfs/zpool refactor

sjorge opened the pull request:
### What does this PR do?
This PR attempts to cleanup the code and reduce complexity for the zfs and zpool modules and states.

- [x] salt.utils.zfs
  - [x]  proper conversion from zfs 'values' to pythonic values
  - [x] command building with proper escaping and conversion for values
  - [x] common code like testing for feature flags, zfs support,...
- [x] salt.grains.zfs
  - [x] use __utils__['zfs.is_supported']
  - [x] use __utils__['zfs.has_feature_flags']
- [x] salt.modules.zpool
  - [x] switch to ```__utils__['zfs.zpool_command']```
  - [x] simply error handing (pass along what zpool complains about)
  - [x] refactor or clarify parsing code
  - [x] improve tests
- [x] salt.modules.zfs
  - [x] switch to ```__utils__['zfs.zfs_command']```
  - [x] simply error handing (pass along what zfs complains about)
  - [x] refactor or clarify parsing code
  - [x] improve tests
- [x] salt.states.zpool
  - [x] update to work with update salt.modules.zpool
  - [x] add tests for salt.states.zpool
- [x] salt.states.zfs
  - [x] update to work with update salt.modules.zfs
  - [x] add tests for salt.states.zfs
- [x] final pass of minor cleanup

### What issues does this PR fix or reference?
#44801

### Previous Behavior
Lots of duplicate code, inconsistencies in output, ...

### New Behavior
Most duplicate code moved to a salt.utils.zfs, only do minimal error handling in modules (zfs/zpool command are pretty good at reporting what is wrong, so we just show that instead), more readable parsing code (hopefully).

Support for ZFS on OSX also added.

### Tests written?
Yes

### Commits signed with GPG?
No

sjorge committed:
Phase 0 - salt.utils.zfs tests

sjorge committed:
Phase 0 - salt.utils.zfs

sjorge committed:
Phase 1 - salt.grains.zfs

sjorge committed:
Phase 1 - salt.modules.zfs tests

sjorge committed:
Phase 1 - salt.modules.zfs

sjorge committed:
Phase 1 - salt.modules.zpool tests

sjorge committed:
Phase 1 - salt.modules.zpool

sjorge commented:
Alright it's back, for some reason every big rebase that fails corrupts my entire repo where it is complaining about weird stuff. Hopefully I will have some time during cfgmgmtcamp to work on this and get it merge. Lots of stuff still to do though.

sjorge committed:
Phase 2 - salt.states.zpool

They old way of passing the zpool layout was a pain to work with.
A new layout has been introduce and the documentation updated accordingly.

The legacy format is still supported.

sjorge committed:
Phase 2 - salt.states.zpool tests

sjorge committed:
Phase 2 - salt.states.zfs

Ch3LL commented:
Go Go Jenkins!

sjorge committed:
Phase 2 - salt.states.zfs tests

sjorge committed:
Phase 3 - Minor cleanups

sjorge committed:
Phase 3 - Cleanup tests

sjorge commented:
@Ch3LL this should be good to merge now. All failing tests are not mine.

Was there any documentation on the new test thingy Bo mentioned? These tests still use the old mocking stuff. I'd like to eventually rewrite them but that won't be for a while. Don't want to poke this code again for at least a few months :)

sjorge commented:
@Ch3LL bump

sjorge commented:
@rallytime can you have a look at this? Looks like Megan is a bit busy atm.
All test failures are not related to my code as far as I can see. I just merged develop into this again.

Ch3LL commented:
apologies @sjorge on the late reply. There is not any documentation currently, but the thing he was referring to was pytest, which you can google on some docs. We will need to write up some more docs for this migration to pytest and make a hardcut for not allowing unittest, but we have not gotten to that yet. Will try to get to it after this oxygen release :) Looks like your tests are still running right now so i'll check back here later today to see if any of the tests are related

rallytime commented on salt/modules/zpool.py:
```diff
@@ -907,12 +925,7 @@ def split(zpool, newzpool, **kwargs):
-
-        salt '*' zpool.split datamirror databackup
-        salt '*' zpool.split datamirror databackup altroot=/backup
+    .. versionadded:: Oxygen
```
I know this wasn't your code, but can you change this to 2018.3.0? We _just_ updated these references in the oxygen branch. :)

sjorge replied:
Done, updates the references in all files I touched.

rallytime commented on salt/states/zfs.py:
```diff
@@ -341,104 +394,194 @@ def filesystem_present(name, create_parent=False, properties=None, cloned_from=N
+def filesystem_present(name, create_parent=False, properties=None, cloned_from=None):
+    '''
+    ensure filesystem exists and has properties set
+
```
This function could use a `.. versionadded:: Fluorine`

sjorge replied:
Why? This fucntions has been around since way before Flourine?

rallytime replied:
Ah, sorry. The way the diff came through made it look like this was a new function.

rallytime reviewed (changes requested):
@sjorge Overall, this is really well done. It will be nice to have all of this more centralized. I just have a couple of smaller comments. If you could fix those up and answer some of my questions, that would be great.

sjorge committed:
Phase 3 - Minor cleanup based on feedback

sjorge commented on salt/grains/zfs.py:
```diff
@@ -16,25 +16,25 @@
 __salt__ = {
     'cmd.run': salt.modules.cmdmod.run,
 }
+__utils__ = {
```
```__utils__``` does not seem to be available.

```
[CRITICAL] Failed to load grains defined in grain file zfs.zfs in function <function zfs at 0x10cf09500>, error:
Traceback (most recent call last):
  File "/Users/sjorge/Desktop/_work/salt_dev/lib/python2.7/site-packages/salt/loader.py", line 755, in grains
    ret = funcs[key]()
  File "/Users/sjorge/Desktop/_work/salt_dev/lib/python2.7/site-packages/salt/grains/zfs.py", line 78, in zfs
    grains['zfs_support'] = __utils__['zfs.is_supported']()
NameError: global name '__utils__' is not defined
```

Is the result without seeding ```__utils__``` manually.

rallytime replied:
🤔 Interesting... @gtmanfred Is `__utils__` not loaded yet here or something?

gtmanfred replied:
Grains has no dunders passed into it.  This is why we have to set `__salt__` manually as well.

rallytime replied:
Ah, OK. Thanks!

sjorge commented on salt/grains/zfs.py:
```diff
@@ -16,25 +16,25 @@
 
 # Import salt libs
-import salt.utils.dictupdate
+import salt.utils.dictupdate as dictupdate
```
Updated, commit incoming later.

sjorge commented on salt/modules/zfs.py:
```diff
@@ -362,164 +293,229 @@ def list_(name=None, **kwargs):
+        properties = properties.split(',')
+
+    # NOTE: name should be first property
+    while 'name' in properties:
```
Because it is possible to enter 'name' multiple times. So we remove all of them.

```
Python 2.7.10 (default, Jul 15 2017, 17:16:57)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> test = ['hello', 'name', 'world', 'name', 'cookies']
>>> if 'name' in test:
...     test.remove('name')
...
>>> print test
['hello', 'world', 'name', 'cookies']
>>> test = ['hello', 'name', 'world', 'name', 'cookies']
>>> while 'name' in test:
...     test.remove('name')
...
>>> print test
['hello', 'world', 'cookies']
>>>
```

I have update the note to make this clear why we do this.

sjorge commented on salt/modules/zfs.py:
```diff
@@ -362,164 +293,229 @@ def list_(name=None, **kwargs):
+        opts['-d'] = kwargs.get('depth')
+    if kwargs.get('type', False):
+        opts['-t'] = kwargs.get('type')
+    if kwargs.get('sort', False) and kwargs.get('sort') in properties:
```
This was done to be consistent with the other around it, the old code used to stuff them all into variables first. It did not make it a lot more readable so started just grabbing them with kwargs.get in the refactor.

But yes, sort was called a lot. I now grab it once and store it in kwargs_sort. Commit coming soon to address this.

sjorge commented on salt/modules/zfs.py:
```diff
@@ -362,164 +293,229 @@ def list_(name=None, **kwargs):
+
+    .. warning::
+
+            Passing '-a' as name is deprecated and will be removed 2 verions after Flourine.
```
Fixed, commit incoming. I also clarified the note om how the if statement should be updated when the time comes to avoid any confusion. I did not know about warn_until, this is handy.

sjorge commented on salt/modules/zfs.py:
```diff
@@ -997,66 +984,55 @@ def release(tag, *snapshot, **kwargs):
-                ret[csnap][ctag] = 'released'
+    '''
+    ## warn about tag change
+    # NOTE: remove me 2 versions after Flourine
```
Fixed

sjorge commented:
@rallytime thans for going over this massive change!
I fixed the things you mentioned aside from two which were done so deliberately. 

For one I updated the note above it to make it clear why this was a while and not an if. For the ```__utils__``` one I left a comment on your comment.

sjorge commented on salt/modules/zfs.py:
```diff
@@ -362,164 +293,229 @@ def list_(name=None, **kwargs):
+
+    .. warning::
+
+            Passing '-a' as name is deprecated and will be removed 2 verions after Flourine.
```
Fixed
